### DISCUSSION TITLE
Return multiple error messages from validate function

### DISCUSSION THREAD

**Discussion by damien-duignan** [react-hook-form/react-hook-form#8362] [Q&A]
I'm trying to write a validate function which returns multiple error messages. I can see from the typescript definitions, that it's possible to pass an array of strings. However, this just returns an empty error message for the input.

Is this the expected behaviour? [Codesandbox example](https://codesandbox.io/s/return-array-from-validate-function-j4obrc).

For background, I'm trying to integrate an external validation library(ajv). I don't know how many errors will be returned, so I can't write the validation property as an object up front. I'd like to avoid writing a resolver, as this will mean that all built-in validators won't work anymore.

Can anyone suggest a workaround for this?

**Comment by bluebill1049** [comment]
can you do try?

```tsx
valdiate: {
  test1: () => {},
  test2: () => {},
}
```

**Reply by damien-duignan** [reply]
Thanks for your reply.

I can't write the object up front as I'm using an external validation library. So I have to write using the callback function instead of an object of callbacks.

So I thought I could do something like this:
```ts
validate: (value) => {
  const errors: { id: string; errorMessage: string }[] = functionToValidateFromExternalLibrary(value);
  return errors.map((error) => error.errorMessage);
}
```

I assumed that I could return an array of strings from this function, since the type `ValidateResult` accepts an array of strings. 
Here's the type definition defined for `ValidateResult`:
```ts
export declare type ValidateResult = Message | Message[] | boolean | undefined;
```

I checked out the function [getValidateError](https://github.com/react-hook-form/react-hook-form/blob/master/src/logic/getValidateError.ts#L17), and it seems that the `message` is only returned when `result` is a string. Passing an array type will just result in `message` being an empty string.

So is there a reason that the validate function accepts an array of strings but doesn't use them? I would've imagined that passing an array of strings would return the errors in the `types` property. Something like this(modified version of `getValidateError`):
```ts
return {
  type,
  message: isMessage(result) ? result : '',
  ref,
  types: Array.isArray(result)
    ? result?.reduce(
        (prev, message, i) => ({
          ...prev,
          [`error${i}`]: message,
        }),
        {},
      )
    : undefined,
};
```

**Reply by damien-duignan** [reply]
Due to this limitation, we'll have to resort to using a Resolver. The Resolver approach isn't ideal as we'll have to rewrite all built-in validations (max, min etc.). This could lead to discrepancies between how `react-hook-form` handles these built-in validators, and how we have implemented them in our own resolver.

We will use some forms with the custom resolver, and some without (so using built-in validators). So we would like to avoid duplicating validation code(and also differences in validation logic).

So I think one of the following approaches would work:

1. The [validation logic](https://github.com/react-hook-form/react-hook-form/blob/master/src/logic/validateField.ts) from `react-hook-form` is exported as a reusable function & we can use it in our custom Resolver.
2. The `validation` function can accept a return value of object, with each value being added to the errors of the input:
```
    validate: (value) => {
      const errors = validateWithExternalLibrary(value);
      /**
       * errors:
       * {
       *  error1: 'This is error 1',
       *  error2: 'This is error 2'
       * }
       */
      return errors;
    },
```

Any thoughts on the proposals?
